File: kds/src/main/java/com/ernesto/kds/data/MerchantFirestore.kt
package com.ernesto.kds.data

import com.google.firebase.firestore.CollectionReference
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.FirebaseFirestore

/**
 * Single source of truth for merchant-scoped Firestore paths in the KDS app.
 *
 * All merchant data lives under `Merchants/{merchantId}/<subcollection>`.
 * Call [init] after pairing resolves the merchant and use [col] / [doc]
 * everywhere instead of raw `db.collection("...")`.
 */
object MerchantFirestore {

    private val db: FirebaseFirestore get() = FirebaseFirestore.getInstance()

    @Volatile
    var merchantId: String = ""
        private set

    val isInitialized: Boolean get() = merchantId.isNotBlank()

    private val NAME_MAP: Map<String, String> = mapOf(
        "Orders" to "orders",
        "Categories" to "categories",
        "subcategories" to "subcategories",
        "MenuItems" to "menuItems",
        "ModifierGroups" to "modifierGroups",
        "ModifierOptions" to "modifierOptions",
        "ItemModifierGroups" to "itemModifierGroups",
        "menus" to "menus",
        "menuSchedules" to "menuSchedules",
        "Employees" to "employees",
        "Customers" to "customers",
        "Taxes" to "taxes",
        "discounts" to "discounts",
        "Printers" to "printers",
        "Batches" to "batches",
        "Transactions" to "transactions",
        "cashLogs" to "cashLogs",
        "kds_devices" to "kds_devices",
        "payment_terminals" to "payment_terminals",
        "Sections" to "sections",
        "tableLayouts" to "tableLayouts",
        "Tables" to "tables",
        "PosDevices" to "posDevices",
        "DeviceActivations" to "deviceActivations",
        "Reservations" to "reservations",
        "Settings" to "settings",
        "settings" to "settings",
        "OnlineHeroSlides" to "onlineHeroSlides",
        "RemotePaymentCommands" to "remotePaymentCommands",
        "OnlineTerminalPaymentRequests" to "onlineTerminalPaymentRequests",
        "Counters" to "counters",
    )

    fun init(merchantId: String) {
        require(merchantId.isNotBlank()) { "merchantId must not be blank" }
        this.merchantId = merchantId
    }

    fun reset() {
        merchantId = ""
    }

    private fun merchantDoc(): DocumentReference =
        db.collection("Merchants").document(merchantId)

    fun resolvedName(name: String): String =
        NAME_MAP[name] ?: name

    fun col(name: String): CollectionReference {
        check(isInitialized) { "MerchantFirestore.init() not called" }
        return merchantDoc().collection(resolvedName(name))
    }

    fun doc(collectionName: String, docId: String): DocumentReference {
        return col(collectionName).document(docId)
    }
}

File: maxipay-dashboard/src/lib/merchantFirestore.ts
import {
  collection,
  doc,
  type CollectionReference,
  type DocumentReference,
} from "firebase/firestore";
import { db } from "@/firebase/firebaseConfig";

/**
 * Old root-level name -> new subcollection name under `Merchants/{mid}/`.
 */
const NAME_MAP: Record<string, string> = {
  Orders: "orders",
  Categories: "categories",
  subcategories: "subcategories",
  MenuItems: "menuItems",
  ModifierGroups: "modifierGroups",
  ModifierOptions: "modifierOptions",
  ItemModifierGroups: "itemModifierGroups",
  menus: "menus",
  menuSchedules: "menuSchedules",
  Employees: "employees",
  Customers: "customers",
  Taxes: "taxes",
  discounts: "discounts",
  Printers: "printers",
  Batches: "batches",
  Transactions: "transactions",
  cashLogs: "cashLogs",
  kds_devices: "kds_devices",
  payment_terminals: "payment_terminals",
  Sections: "sections",
  tableLayouts: "tableLayouts",
  Tables: "tables",
  PosDevices: "posDevices",
  DeviceActivations: "deviceActivations",
  Reservations: "reservations",
  Settings: "settings",
  settings: "settings",
  OnlineHeroSlides: "onlineHeroSlides",
  RemotePaymentCommands: "remotePaymentCommands",
  OnlineTerminalPaymentRequests: "onlineTerminalPaymentRequests",
  Counters: "counters",
};

export function resolvedName(name: string): string {
  return NAME_MAP[name] ?? name;
}

/**
 * Non-empty merchant document id under `Merchants/{id}/`.
 * Prevents invalid paths like `Merchants/posDevices/...` when [merchantId] is blank
 * (e.g. stale closures or missing JWT `merchantId` after admin creates the store).
 */
function requireMerchantId(merchantId: string): string {
  const mid = merchantId.trim();
  if (!mid) {
    throw new Error(
      "Missing merchant id for this Firestore path. Refresh the page or sign out and back in. New stores need a login that includes merchantId on the token (set when the merchant is created in admin).",
    );
  }
  return mid;
}

/**
 * Merchant-scoped collection: `Merchants/{merchantId}/{resolvedName}`.
 * Accepts either old root name (e.g. `"Orders"`) or new name (`"orders"`).
 */
export function merchantCol(
  merchantId: string,
  name: string
): CollectionReference {
  const mid = requireMerchantId(merchantId);
  return collection(db, "Merchants", mid, resolvedName(name));
}

/**
 * Merchant-scoped document: `Merchants/{merchantId}/{resolvedName}/{docId}`.
 */
export function merchantDoc(
  merchantId: string,
  collectionName: string,
  docId: string
): DocumentReference {
  const mid = requireMerchantId(merchantId);
  return doc(db, "Merchants", mid, resolvedName(collectionName), docId);
}
